modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiBackstress.C

```diff
@@ -0,0 +1,207 @@
+    if (effective_stress < 0.0)
+      _slip_increment[_qp][i] *= -1.0;
+
+    if (std::abs(_slip_increment[_qp][i]) * _substep_dt > _slip_incr_tol)
+    {
+      if (_print_convergence_message)
+        mooseWarning("Maximum allowable slip increment exceeded ",
+                     std::abs(_slip_increment[_qp][i]) * _substep_dt);
+
+      return false;
+    }
+  }
+  return true;
+}
+
+void
+CrystalPlasticityKalidindiBackstress::calculateConstitutiveSlipDerivative(
+    std::vector<Real> & dslip_dtau)
+{
+  Real effective_stress;
+
+  for (const auto i : make_range(_number_slip_systems))
+  {
+    effective_stress = _tau[_qp][i] - _backstress[_qp][i];
+
+    if (MooseUtils::absoluteFuzzyEqual(effective_stress, 0.0))
+      dslip_dtau[i] = 0.0;
+    else
+      dslip_dtau[i] = _ao / _xm *
+                      std::pow(std::abs(effective_stress / _slip_resistance[_qp][i]), 1.0 / _xm - 1.0) /
+                      _slip_resistance[_qp][i];
+  }
+}
+
+bool
+CrystalPlasticityKalidindiBackstress::areConstitutiveStateVariablesConverged()
+{
+  return isConstitutiveStateVariableConverged(_slip_resistance[_qp],
+                                              _slip_resistance_before_update,
+                                              _previous_substep_slip_resistance,
+                                              _resistance_tol);                                        
+}
+
+void
+CrystalPlasticityKalidindiBackstress::updateSubstepConstitutiveVariableValues()
+{
+  _previous_substep_slip_resistance = _slip_resistance[_qp];
+  _previous_substep_backstress = _backstress[_qp];
+}
+
+void
+CrystalPlasticityKalidindiBackstress::cacheStateVariablesBeforeUpdate()
+{
+  _slip_resistance_before_update = _slip_resistance[_qp];
+  _backstress_before_update = _backstress[_qp];
+}
+
+void
+CrystalPlasticityKalidindiBackstress::calculateStateVariableEvolutionRateComponent()
+{
+  CrystalPlasticityKalidindiUpdate::calculateStateVariableEvolutionRateComponent();
+  
+  ArmstrongFrederickBackstressUpdate();
+}
+
+bool
+CrystalPlasticityKalidindiBackstress::updateStateVariables()
+{
+  // Now perform the check to see if the slip system should be updated
+  for (const auto i : make_range(_number_slip_systems))
+  {
+    _slip_resistance_increment[i] *= _substep_dt;
+    if (_previous_substep_slip_resistance[i] < _zero_tol && _slip_resistance_increment[i] < 0.0)
+      _slip_resistance[_qp][i] = _previous_substep_slip_resistance[i];
+    else
+      _slip_resistance[_qp][i] =
+          _previous_substep_slip_resistance[i] + _slip_resistance_increment[i];
+
+    if (_slip_resistance[_qp][i] < 0.0)
+      return false;
+  }
+
+  // Backstress: can be both positive or negative
+  for (const auto i : make_range(_number_slip_systems))
+  { 
+    _backstress_increment[i] *= _substep_dt;
+    _backstress[_qp][i] = _previous_substep_backstress[i] + _backstress_increment[i];
+  }
+
+  return true;
+}
```

Comment:
```suggestion
  // Now perform the check to see if the slip system should be updated
  if (CrystalPlasticityKalidindiUpdate::updateStateVariables())
  {
    // now update backstress variable, which can be positive or negative
    for (const auto i : make_range(_number_slip_systems))
    { 
      _backstress_increment[i] *= _substep_dt;
      _backstress[_qp][i] = _previous_substep_backstress[i] + _backstress_increment[i];
    }
    return true;
  }
  else
    return false;
}
```

Reply:
This suggested change avoids duplicating the code in the `CrystalPlasticityKalidindiUpdate` class

Reply:
resolved